r/R/dplyr-distinct.R

```diff
@@ -0,0 +1,71 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# The following S3 methods are registered on load if dplyr is present
+
+distinct.arrow_dplyr_query <- function(.data, ..., .keep_all = FALSE) {
+
+  if (.keep_all == TRUE) {
+    # After ARROW-13767 is merged, we can implement this via e.g.
+    # iris %>% group_by(Species) %>% slice(1) %>% ungroup()
```

Comment:
I'm not sure this is accurate

Reply:
I don't think ARROW-13767 will help us implement `.keep_all = TRUE`. To implement that, I think we would need a hash aggregate function `hash_first()` that returns the first value of a column within each group. However since row order within groups is potentially non-deterministic, this is perhaps better understood as a function that just grabs any one column value from within the group.

Reply:
I opened ARROW-13993 for a hash aggregate function that returns the first value (effectively any one arbitrary value) from within each group

Reply:
Thanks @ianmcook !

Reply:
@thisisnic could you please open a separate Jira to implement `.keep_all = TRUE` for 7.0.0? Thanks